From 4fde022f7702bca41474038aeea5d0088cb991f3 Mon Sep 17 00:00:00 2001
From: Jianqiang Chen <jianqiang.chen@xilinx.com>
Date: Tue, 9 Mar 2021 12:36:38 -0800
Subject: [PATCH 1767/1852] drm: xlnx: zynqmp_disp: Move live bridges
 registration to probe

commit 224391f8d756889166bcf282e897a23267290825 from
https://github.com/Xilinx/linux-xlnx.git

Currently the live bridges are registered during bind because it
expects the PS DP CRTC to be used. When we use an external CRTC, this
causes a dead lock because the external CRTC requires the live bridges
to be initialized before it can probe successfully. Meanwhile binding
requires all components to be probed first. This patch registers the
live bridges during probe to avoid this deadlock.

Signed-off-by: Jianqiang Chen <jianqiang.chen@xilinx.com>
Signed-off-by: Dylan Yip <dylan.yip@xilinx.com>
Acked-by: Varunkumar Allagadapa <varunkumar.allagadapa@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/zynqmp_disp.c | 28 +++++++++++++++++-----------
 1 file changed, 17 insertions(+), 11 deletions(-)

diff --git a/drivers/gpu/drm/xlnx/zynqmp_disp.c b/drivers/gpu/drm/xlnx/zynqmp_disp.c
index b3ba23c34d8f..2fd0b71ded71 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_disp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_disp.c
@@ -2792,17 +2792,6 @@ static int zynqmp_disp_create_plane(struct zynqmp_disp *disp)
 		type = DRM_PLANE_TYPE_PRIMARY;
 	}
 
-	for (i = 0; i < ZYNQMP_DISP_NUM_LAYERS; i++) {
-		layer = &disp->layers[i];
-		layer->bridge.enable = &zynqmp_disp_bridge_enable;
-		layer->bridge.disable = &zynqmp_disp_bridge_disable;
-		layer->bridge.set_input = &zynqmp_disp_bridge_set_input;
-		layer->bridge.get_input_fmts =
-			&zynqmp_disp_bridge_get_input_fmts;
-		layer->bridge.of_node = layer->of_node;
-		xlnx_bridge_register(&layer->bridge);
-	}
-
 	/* Attach properties to each layers */
 	drm_object_attach_property(&layer->plane.base, disp->g_alpha_prop,
 				   ZYNQMP_DISP_V_BLEND_SET_GLOBAL_ALPHA_MAX);
@@ -3239,6 +3228,8 @@ int zynqmp_disp_probe(struct platform_device *pdev)
 	struct zynqmp_disp *disp;
 	struct resource *res;
 	int ret;
+	struct zynqmp_disp_layer *layer;
+	unsigned int i;
 
 	disp = devm_kzalloc(&pdev->dev, sizeof(*disp), GFP_KERNEL);
 	if (!disp)
@@ -3338,6 +3329,21 @@ int zynqmp_disp_probe(struct platform_device *pdev)
 
 	zynqmp_disp_init(disp);
 
+	/*
+	 * Register live bridges so external CRTCs will be able probe
+	 * successfully
+	 */
+	for (i = 0; i < ZYNQMP_DISP_NUM_LAYERS; i++) {
+		layer = &disp->layers[i];
+		layer->bridge.enable = &zynqmp_disp_bridge_enable;
+		layer->bridge.disable = &zynqmp_disp_bridge_disable;
+		layer->bridge.set_input = &zynqmp_disp_bridge_set_input;
+		layer->bridge.get_input_fmts =
+			&zynqmp_disp_bridge_get_input_fmts;
+		layer->bridge.of_node = layer->of_node;
+		xlnx_bridge_register(&layer->bridge);
+	}
+
 	return 0;
 
 error_aclk:
-- 
2.31.1

